tools/src/main/java/org/apache/kafka/tools/EndToEndLatency.java

```diff
@@ -0,0 +1,237 @@
+ * num_messages = # messages to send
+ * producer_acks = See ProducerConfig.ACKS_DOC
+ * message_size_bytes = size of each message in bytes
+ *
+ * e.g. [localhost:9092 test 10000 1 20]
+ */
+public class EndToEndLatency {
+
+    private final static long POLL_TIMEOUT_MS = 60000;
+
+    public static void main(String... args) {
+        Exit.exit(mainNoExit(args));
+    }
+
+    static int mainNoExit(String... args) {
+        try {
+            if (args.length != 5 && args.length != 6) {
+                System.err.println("USAGE: java " + EndToEndLatency.class.getName()
+                        + " broker_list topic num_messages producer_acks message_size_bytes [optional] properties_file");
+                return 1;
+            }
+            execute(args);
+            return 0;
+        } catch (TerseException e) {
+            System.err.println(e.getMessage());
+            return 1;
+        } catch (Throwable e) {
+            System.err.println(e.getMessage());
+            System.err.println(Utils.stackTrace(e));
+            return 1;
+        }
+    }
+
+    // Visible for testing
+    static void execute(String... args) throws Exception {
+        String brokers = args[0];
+        String topic = args[1];
+        int numMessages = Integer.parseInt(args[2]);
+        String acks = args[3];
+        int messageSizeBytes = Integer.parseInt(args[4]);
+        String propertiesFile = args.length > 5 ? args[5] : null;
+
+        if (!Arrays.asList("1", "all").contains(acks)) {
+            throw new IllegalArgumentException("Latency testing requires synchronous acknowledgement. Please use 1 or all");
+        }
+
+        Properties props;
+        try {
+            props = loadPropsWithBootstrapServers(propertiesFile);
+        } catch (FileNotFoundException e) {
+            throw new IllegalArgumentException("Properties file not found");
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+
+        try (KafkaConsumer<byte[], byte[]> consumer = createKafkaConsumer(props, brokers);
+             KafkaProducer<byte[], byte[]> producer = createKafkaProducer(props, acks, brokers)) {
+
+            if (!consumer.listTopics().containsKey(topic)) {
+                createTopic(props, topic);
+            }
+
+            setupConsumer(topic, consumer);
+
+            double totalTime = 0.0;
+            long[] latencies = new long[numMessages];
+            Random random = new Random(0);
+
+            for (int i = 0; i < numMessages; i++) {
+                byte[] message = randomBytesOfLen(random, messageSizeBytes);
+                long begin = System.nanoTime();
+                //Send message (of random bytes) synchronously then immediately poll for it
+                producer.send(new ProducerRecord<>(topic, message)).get();
+                ConsumerRecords<byte[], byte[]> records = consumer.poll(Duration.ofMillis(POLL_TIMEOUT_MS));
```

Comment:
The reason I am not using the iterator is that `ConsumerRecord` exposes functions like `isEmpty` and `count` which seemed easier to understand when used in `validate` method. Let me know if you think otherwise.

Reply:
I would prefer to stick with the original code.

Reply:
I don't think we need to keep the original code when it's strictly worse.

Reply:
Yeah in this case I feel using the methods provided by `ConsumerRecords` seems cleaner. 

> We should keep the original code if there isn't a clear improvement from changing and whatever changes we do should be localized - changes that affect many methods, other files, etc. are best avoided if possible.

Ack. Would keep that in mind.

Reply:
> I don't think we need to keep the original code when it's strictly worse (for example the original code ends up exhausting the iterator to get the size).
> 

Fair enough. Thanks.